.changeset/gold-badgers-float.md

```diff
@@ -0,0 +1,23 @@
+---
+"@fluidframework/shared-object-base": minor
+---
+
+Added typeguard for SharedObjectKinds
+
+In the 2.0 release of Fluid, the concrete class implementations for DDSes were hidden from Fluid's API surface.
+This made `instanceof` checks fail to work correctly.
+There were ways to work around this in application code, but they involved boilerplate which required more understanding of Fluid internals than should be necessary.
+This change adds a drop-in replacement to `instanceof`: the `.is()` method to `SharedObjectKind`.
+For example:
+
+```typescript
+// Works in FluidFramework 1.0 but not in the initial release of FluidFramework 2.0:
+if (myObject instanceof SharedString) {
+	// do something
+}
+
+// With this change, that code can now be written like so:
+if (SharedString.is(myObject)) {
```

Comment:
We could just "name" the type guard Symbol.hasInstance, then instancof will work again.

Generally "is" makes sense as a narrowing method on the thing being narrow. Ex A.is(B) tells you if A is a B. THis case is backwards, and I think instancof / Symbol.hasInstance may make more sense, and better align with language/ecosystem conventions.

Reply:
+1 that it reads backwards/confusingly

Reply:
yeah I thought the same thing when writing this.

I'm on board with the suggestion here.

Reply:
hm. I think this might not be feasible until we adopt [TS 5.5](https://github.com/microsoft/TypeScript/commit/ea0d228f83adba354d6830e559421bc5a60e0d72) (and probably raise min support to 5.5) due to https://github.com/microsoft/TypeScript/issues/56536.

Reply:
we could have both with a remarks comment documenting this, and I think apps that are already on that version of TS would be able to use instanceof just fine.

Reply:
I though we decided we didn't like instanceof overriding as its generally not a good practice for types your don't fully control, and not a pattern we want to encourage fluid customers to us

Reply:
in terms of `is` being backwards, i don't disagree, but i also like how short and concise it is, and i don't think it's really that confusing. Do anyone have ideas for an alternate naming?

Reply:
I think I agree with Tony here. No other meaning really makes sense for a static method, and the docs clarify what it does. I expect any confusion would be _very_ short.

We could name it `describes` or `describesInstance` or something, but I don't think something of that form is much easier to understand, and is definitely clunkier.

Reply:
Just my 2 cents - this seems like a fine occasion for `instanceof`, but if we don't/can't support that, then I'd prefer either of the following to the backwards `SharedString.is`:

* `SharedString.isSharedString(value)`
* `isSharedString(value)`   (free function)

Reply:
We talked this over more in API council and have decided to proceed with `.is` for now. We can revisit support for `instanceof` in the future if it's a problem.

Reply:
Some elaboration on the rationale:

- We prefer centralizing the logic rather than export free functions / constants / whatever in each package due to the maintenance burden. This also aligns with the point Craig wrote: "One of the reasons I created SharedObjectKind was to be able to add things like this to all of them in a consistent way easily." Even if we dedupe the implementation, adding `isSharedString` and friends is still 2 lines per DDS (index export + something like `const isSharedString = createSharedObjectTypeguard(SharedString);`). That alternative will also be larger bundle size as soon as a client uses multiple shared objects' type guards.

Because `instanceof` won't work for some consumers within our support matrix due to TS bug I linked earlier in this section, only having that is a nonstarter.

We could have implemented both `Symbol.hasInstance` and kept `.is` or equivalent around until it was no longer necessary. [This comment](https://github.com/microsoft/FluidFramework/pull/21850#discussion_r1676430083) and concerns around encouraging `instaceof` on Fluid objects made us decide against it--at least for now.

re: naming and confusion there, nobody came up with anything better, and it roughly aligns with some other existing APIs we have (`Tree.is`, `Marker.is`).